crates/ruff/src/rules/flake8_pyi/rules/custom_typevar_return.rs

```diff
@@ -0,0 +1,203 @@
+///         ...
+///
+///     def foo(self: Self, arg: bytes) -> Self:
+///         ...
+///
+///     @classmethod
+///     def bar(cls: type[Self], arg: int) -> Self:
+///         ...
+/// ```
+#[violation]
+pub struct CustomTypeVarReturnType {
+    method_name: String,
+}
+
+impl Violation for CustomTypeVarReturnType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let CustomTypeVarReturnType { method_name } = self;
+        format!(
+            "Methods like {method_name} should return `typing.Self` instead of a custom TypeVar"
+        )
+    }
+}
+
+/// PYI019
+pub(crate) fn custom_typevar_return_type(
+    checker: &mut Checker,
+    name: &str,
+    decorator_list: &[Decorator],
+    returns: Option<&Expr>,
+    args: &Arguments,
+    type_params: &[TypeParam],
+) {
+    let ScopeKind::Class(_) = checker.semantic().scope().kind else {
+        return;
+    };
+
+    if args.args.is_empty() && args.posonlyargs.is_empty() {
+        return;
+    }
+
+    let Some(returns) = returns else {
+        return;
+    };
+
+    let return_annotation = if let Expr::Subscript(ast::ExprSubscript { value, .. }) = returns {
+        // Ex) `Type[T]`
+        value
+    } else {
+        // Ex) `Type`, `typing.Type`
+        returns
+    };
+
+    // Skip any abstract, static and overloaded methods.
+    if is_abstract(decorator_list, checker.semantic())
+        || is_overload(decorator_list, checker.semantic())
+        || is_staticmethod(decorator_list, checker.semantic())
+    {
+        return;
+    }
+
+    let is_violation: bool =
+        if is_classmethod(decorator_list, checker.semantic()) || name == "__new__" {
+            check_class_method_for_bad_typevars(checker, args, return_annotation, type_params)
+        } else {
+            // If not static, or a class method or __new__ we know it is an instance method
+            check_instance_method_for_bad_typevars(checker, args, return_annotation, type_params)
+        };
+
+    if is_violation {
+        checker.diagnostics.push(Diagnostic::new(
+            CustomTypeVarReturnType {
+                method_name: name.to_string(),
+            },
+            return_annotation.range(),
+        ));
+    }
+}
+
+fn check_class_method_for_bad_typevars(
+    checker: &Checker,
+    args: &Arguments,
+    return_annotation: &Expr,
+    type_params: &[TypeParam],
+) -> bool {
+    let ArgWithDefault { def, .. } = &args.args[0];
+
+    let Some(annotation) = &def.annotation else {
+        return false
+    };
+
+    let Expr::Subscript(ast::ExprSubscript{slice, value, ..}) = annotation.as_ref() else {
+        return false
+    };
+
+    let Expr::Name(ast::ExprName{ id: id_value, .. }) = value.as_ref() else {
+        return false
+    };
+
+    // Don't error if the first argument is annotated with typing.Type[T]
+    // These are edge cases, and it's hard to give good error messages for them.
+    if id_value != "type" {
+        return false;
+    };
```

Comment:
Upstream implementation mentions the exception as follows:

```python
        # Don't error if the first argument is annotated
        # with `builtins.type[T]` or `typing.Type[T]`
        # These are edge cases, and it's hard to give good error messages for them.
        if not _is_name(first_arg_annotation.value, "type"):
            return
```
Source: https://github.com/PyCQA/flake8-pyi/blob/2a86db8271dc500247a8a69419536240334731cf/pyi.py#L1843C1-L1847 

Unsure if this was a typo or I am missing something. But it seems to me that `builtins.type[T]` is not the edge case but the thing we want to check. 

I've implemented what I think they meant. But just mentioning it here as I might be confused or missing something.

Reply:
I think what I was trying to say in that comment was that we emit Y019 for this:

```python
def __new__(cls: type[S]) -> S: ...
```

But not for any of these:

```python
def __new__(cls: builtins.type[S]) -> S: ...
def __new__(cls: Type[S]) -> S: ...
def __new__(cls: typing.Type[S]) -> S: ...
```

Whereas the behaviour most consistent with how flake8-pyi usually works would be to emit the same error for all of them.

I can't promise that this is the optimal  behaviour, though, just what seemed most reasonable to me at the time :D

Reply:
Hey Alex,

Thanks for chipping in! Then the confusion is on me, and not the comment 😅 .

Can you explain why `builtins.type[S]` is more problematic than just `type[S]`? Or `typing.Type` for that matter? Is it just a message formatting thing or are the functional differences?

Reply:
Yeah I think, as the comment says, it just felt complicated to get a good error message in that kind of situation without excessively complicating the code. For `Type` and `typing.Type`, I think I might also have been worried about having two error messages telling the user to do contradictory things (one flake8-pyi code tells you to use `type` instead of `Type`; the other tells you to use `Self`, what do you do?!)

But I can't really remember the exact reasoning for making the decision and it's entirely possible we could/should revisit it at flake8-pyi!

Reply:
Clear! I'll leave it like this for now. Thanks!